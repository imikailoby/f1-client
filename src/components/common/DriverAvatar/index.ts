export { default as DriverAvatar } from './driverAvatar.component';
