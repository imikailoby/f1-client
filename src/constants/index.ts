export const API_URL: string = process.env.REACT_APP_API_URL || 'http://ergast.com/api/f1';
export const CURRENT_ENV: 'development' | 'production' | 'test' = process.env.NODE_ENV as
  | 'development'
  | 'production'
  | 'test';

// Value hardcoded because we cannot time traveling and start first F1 season before 1950;
export const FIRST_F1_SEASON_YEAR = 1950;
export const FROM_F1_SEASON_YEAR = Number(process.env.REACT_APP_FROM_F1_SEASON_YEAR || '2005');

export const WIKI_API_URL = 'https://en.wikipedia.org/w/api.php';
