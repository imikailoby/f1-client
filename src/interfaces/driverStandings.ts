import { Constructor } from './constructor';
import { Driver } from './driver';

export interface DriverStandingsItem {
  Constructors: Constructor[];
  Driver: Driver,
  points: string;
  positionText: string;
  wins: string;
}
