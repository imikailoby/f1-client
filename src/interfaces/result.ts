import { Constructor } from './constructor';
import { Driver } from './driver';

export interface Result {
  number: string;
  position: string;
  positionText: string;
  points: string;
  Driver: Driver;
  Constructor: Constructor;
  grid: string;
  laps: string;
  status: string;
  Time: {
    millis: string;
    time: string;
  };
  FastestLap: {
    rank: string;
    lap: string;
    Time: {
      time: string;
    };
    AverageSpeed: {
      units: string;
      speed: string;
    };
  };
}
