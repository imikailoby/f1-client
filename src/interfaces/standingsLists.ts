import { DriverStandingsItem } from './driverStandings';

export interface StandingsListItem {
  season: string;
  round: string;
  DriverStandings: [DriverStandingsItem];
}
