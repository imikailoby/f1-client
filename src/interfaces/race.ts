import { Result } from './result';

export interface Race {
  season: string;
  round: string;
  url: string;
  raceName: string;
  Circuit: {
    circuitId: string;
    url: string;
    circuitName: string;
    Location: {
      lat: string;
      long: string;
      locality: string;
      country: string;
    };
  };
  date: string;
  time: string;
  Results: [Result];
}
