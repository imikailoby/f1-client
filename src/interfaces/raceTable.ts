import { Race } from './race';

export interface RaceTable {
  season: string;
  position: string;
  Races: Race[];
}
