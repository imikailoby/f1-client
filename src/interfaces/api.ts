import { MRData } from '../types/mrData';
import { RaceTable } from './raceTable';
import { StandingsListItem } from './standingsLists';

export interface IGetDriverStandingsResponse {
  MRData: MRData<{ StandingsTable: { StandingsLists: StandingsListItem[] } }>
}

export interface IGetResultsByYearResponse {
  MRData: MRData<{ RaceTable: RaceTable }>
}

export interface IGetDriverStandings {
  standingsLists: StandingsListItem[],
  pagination: {
    limit: MRData['limit'];
    offset: MRData['offset'];
    total: MRData['total'];
  }
}
