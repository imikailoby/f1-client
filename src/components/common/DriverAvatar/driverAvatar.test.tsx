/**
 * @jest-environment jsdom
 */
import { rest } from 'msw';
import { render } from '@testing-library/react';
import { server } from '../../../../mocks/server';
import getDriverStandingsReponse from '../../../../mocks/responses/getDriverStandingsReponse.json';
import wikipediaApiResponse from '../../../../mocks/responses/wikipediaApiResponse.json';
import { clearSessionStorage } from '../../../utils/common';
import { DriverStandingsItem } from '../../../interfaces/driverStandings';
import { WIKI_API_URL } from '../../../constants';
import { DriverAvatar } from '.';

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => {
  server.close();
  clearSessionStorage();
});

const MOCKED_DS = getDriverStandingsReponse.MRData.StandingsTable.StandingsLists[0]
  .DriverStandings[0] as unknown as DriverStandingsItem;

describe('<DriverAvatar />', () => {
  test('renders with loading state without crashing', () => {
    server.use(rest.get(`${WIKI_API_URL}*`, (_, res, ctx) => res(ctx.status(500))));
    const { container } = render(<DriverAvatar number='??' />);
    expect(container).toMatchSnapshot();
    expect(container.getElementsByTagName('img')).toHaveLength(0);
  });

  test('renders with driver info without crashing', () => {
    server.use(rest.get(`${WIKI_API_URL}*`, (_, res, ctx) => res(ctx.status(200), ctx.json(wikipediaApiResponse))));
    sessionStorage.setItem(MOCKED_DS.Driver.code, 'https://test.com/avatar.png');
    const { container } = render(<DriverAvatar className='test' number='14' driverStanding={MOCKED_DS} />);
    expect(container).toMatchSnapshot();
    expect(container.getElementsByTagName('img')).toHaveLength(1);
  });
});
