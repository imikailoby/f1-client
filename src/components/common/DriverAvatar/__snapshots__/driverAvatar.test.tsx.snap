// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DriverAvatar /> renders with driver info without crashing 1`] = `
<div>
  <div
    class="avatar test"
  >
    <img
      alt="driver avatar"
      loading="lazy"
      src="https://test.com/avatar.png"
    />
    <span
      class="number"
    >
      14
    </span>
  </div>
</div>
`;

exports[`<DriverAvatar /> renders with loading state without crashing 1`] = `
<div>
  <div
    class="avatar undefined"
  >
    <span>
      ...
    </span>
    <span
      class="number"
    >
      ??
    </span>
  </div>
</div>
`;
